File: src/test/java/com/github/vatbub/commandlineUserPromptProcessor/PromptTest.java
package com.github.vatbub.commandlineUserPromptProcessor;

/*-
 * #%L
 * Vatbub Commandline User Prompt Processor
 * %%
 * Copyright (C) 2017 Frederik Kammel
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */


import com.github.vatbub.commandlineUserPromptProcessor.parsables.Parsable;
import com.github.vatbub.commandlineUserPromptProcessor.parsables.ParseException;
import org.junit.Assert;
import org.junit.Rule;
import org.junit.contrib.java.lang.system.SystemOutRule;
import org.junit.contrib.java.lang.system.TextFromStandardInputStream;

/**
 * Created by Frederik on 03.09.2017.
 */
public abstract class PromptTest {
    @Rule
    public final TextFromStandardInputStream systemInMock
            = TextFromStandardInputStream.emptyStandardInputStream();
    @Rule
    public final SystemOutRule systemOutRule = new SystemOutRule().enableLog();

    protected Parsable promptTestCore(String promptText, String input, Parsable returnValue) throws ParseException {
        systemInMock.provideLines(input);
        Prompt prompt = new Prompt(promptText, returnValue);
        returnValue = prompt.doPrompt();

        StringBuilder expectedPromptText = new StringBuilder(promptText);
        String optionsString = returnValue.getOptionsString();
        String defaultValueString = returnValue.getStringForDefaultValue();

        if (returnValue.getDefaultValue() == null || defaultValueString == null) {
            Assert.assertEquals(returnValue.getDefaultValue(), defaultValueString);
        }

        if (optionsString != null) {
            expectedPromptText.append(" [");
            expectedPromptText.append(optionsString);
            expectedPromptText.append("]");
        }

        if (defaultValueString != null) {
            expectedPromptText.append(" [default: ");
            expectedPromptText.append(defaultValueString);
            expectedPromptText.append("]");
        }

        expectedPromptText.append(": ");

        Assert.assertEquals(expectedPromptText.toString(), systemOutRule.getLog().replace("\n", "").replace("\r", ""));
        systemOutRule.clearLog();

        return returnValue;
    }
}

File: src/test/kotlin/com/github/vatbub/commandlineUserPromptProcessor/PromptTest.kt
package com.github.vatbub.commandlineUserPromptProcessor
/*-
 * #%L
 * Vatbub Commandline User Prompt Processor
 * %%
 * Copyright (C) 2017 Frederik Kammel
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import com.github.vatbub.commandlineUserPromptProcessor.parsers.Parser
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.nio.charset.Charset
import java.util.concurrent.TimeUnit.SECONDS

/**
 * Created by Frederik on 03.09.2017.
 */
class PromptTest {

    private fun <T> promptTest(
        promptText: String,
        expectedInput: String,
        expectedDefaultValue: T?,
        mockParsingResult: T,
        expectedOptionsString: String?
    ) {
        var parserToStringCalled = false

        val parser = object : Parser<T> {
            override fun parse(input: String, defaultValue: T?): T {
                assertEquals(expectedInput, input)
                assertEquals(expectedDefaultValue, defaultValue)
                return mockParsingResult
            }

            override fun toString(obj: T): String {
                parserToStringCalled = true
                assertEquals(expectedDefaultValue, obj)
                return super.toString(obj)
            }

            override val optionsString = expectedOptionsString

        }

        val inputStream = ByteArrayInputStream("$expectedInput\n".toByteArray(Charset.forName("UTF-8")))
        val byteArrayOutputStream = ByteArrayOutputStream()
        val printStream = PrintStream(byteArrayOutputStream)

        val prompt = Prompt(promptText, parser)
        assertEquals(
            mockParsingResult,
            prompt.doPrompt(defaultValue = expectedDefaultValue, outStream = printStream, inStream = inputStream)
        )
        if (expectedDefaultValue != null)
            await().atMost(5, SECONDS).until { parserToStringCalled }

        val expectedPromptText = StringBuilder(promptText)
        expectedOptionsString?.let { expectedPromptText.append(" [$it]") }
        expectedDefaultValue?.let { expectedPromptText.append(" [default: $it]") }
        expectedPromptText.append(": ")

        val printedLog = byteArrayOutputStream.toByteArray().toString(Charset.forName("UTF-8"))
        assertEquals(expectedPromptText.toString(), printedLog.replace("\n", "").replace("\r", ""))
    }

    @Test
    fun promptWithDefaultValueTest() = promptTest(
        promptText = "I am a prompt",
        expectedInput = "dummyInput",
        expectedDefaultValue = "I am a default value",
        mockParsingResult = "Parse result",
        expectedOptionsString = "Dummy options"
    )

    @Test
    fun promptNoDefaultValueTest() = promptTest(
        promptText = "I am another prompt",
        expectedInput = "hello world",
        expectedDefaultValue = null,
        mockParsingResult = "Another parse result",
        expectedOptionsString = "And more options"
    )
}